Source/WebCore/css/CSSCounterStyleRegistry.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (C) 2023 Apple Inc.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "CSSCounterStyleRegistry.h"
+
+#include "CSSCounterStyle.h"
+#include "CSSCounterStyleRule.h"
+#include "CSSPrimitiveValue.h"
+#include "CSSValuePair.h"
+
+namespace WebCore {
+
+void CSSCounterStyleRegistry::resolveReferences()
+{
+    // FIXME: does this needs thread-safety guards?
+    if (!hasUnresolvedReferences)
+        return;
+
+    for (auto& [name, counter] : m_authorCounterStyles) {
+        if (counter->isFallbackUnresolved())
+            resolveFallbackReference(*counter);
+        if (counter->isExtendsSystem() && counter->isExtendsUnresolved())
+            resolveExtendsReference(*counter);
+    }
+    hasUnresolvedReferences = false;
+}
+
+void CSSCounterStyleRegistry::resolveExtendsReference(CSSCounterStyle& counterStyle)
+{
+    HashSet<CSSCounterStyle*> countersInChain;
+
+    std::function<void(CSSCounterStyle&)> resolve = [&](CSSCounterStyle& counterStyle) {
```

Comment:
I don't understand why this is a lambda.

Also, `auto` is faster than `std::function` for lambdas.

Reply:
It was auto before but the compiler can't deduce auto for recursive lambdas. I thought it was better than having a
 private member function that would only be used here, because it gets encapsulated, but I can change if you prefer it.

Reply:
Let's just use static or member function if you need recursion until it is in the language (in C++23). std::function introduces an unnecessary virtual function (also we use WTF::Function).